---
title: The `Metacoder` Package
output: 
  rmarkdown::html_vignette:
    toc: false
    css: custom.css
vignette: >
  %\VignetteIndexEntry{Package introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(knitr)
opts_chunk$set(dev='png', fig.width = 10, fig.height = 10, warning = FALSE, message = FALSE,
               autodep = TRUE, cache = TRUE)
```

## An R package for planning amplicon metagenomics research and plotting taxonomy-associated information

*Zachary S. L. Foster and Niklaus J. Gr√ºnwald*

### Introduction 

Metabarcoding is revolutionizing microbial ecology and presenting new challenges:

* Numerous database formats make taxonomic data difficult to parse, combine, and subset.
* Stacked bar charts, commonly used to depict community diversity, lack taxonomic context.
* Barcode loci and primers are a source of under-explored bias.

MetacodeR is an R package that attempts to addresses these issues:

* Sources of taxonomic data can be extracted from any file format. 
* Community diversity can be visualized by color and size in a tree layout.
* Primer specificity can be estimated with *In silico* PCR.

### Extracting taxonomic data

Most databases have a unique file format and taxonomic hierarchy/nomenclature.
Taxonomic data can be extracted from any file format using the **extract_taxonomy** function, as long as one of the following are present: taxon name, taxon ID, sequence ID or embedded classification. 
A regular expression with capture groups and a corresponding key is used to define how to parse the file.
The example code below parses the 16s Ribosome Database Project training set for Mothur:

```{r, warning=FALSE, message=FALSE}
library(metacoder)
seqs <- ape::read.FASTA("trainset10_082014.rdp.fasta")
cat(names(seqs)[1]) # print an example of the sequence headers
data <- extract_taxonomy(seqs, regex = "^(.*)\\t(.*)", key = c("item_info", "class_name"),
                         class_tax_sep = ";", database = "none")
head(taxon_data(data)) # print a few rows of the taxon data
head(item_data(data)) # print a few rows of the sequence data
```

The resulting object contains sequence information associated with an inferred taxonomic hierarchy.

### Metadiversity plots

The hierarchical nature of taxonomic data makes it difficult to plot effectively.
Most often, bar charts, stacked bar charts, or pie graphs are used, but these are ineffective when plotting many taxa or multiple ranks.
The function plot_taxonomy makes a tree utilizing color and size to display taxonomic data (e.g. sequence abundance) that we call a **Metadiversity plot**:

```{r}
plot(data, vertex_size = item_count, 
     vertex_label = name, vertex_color = item_count)
```

Any statistic can be mapped to tree vertex/edge color and size.
The size range displayed is optimized for each graph.
Plotting is flexible enough for pipelines and controlled enough for publication-ready graphics.
There are many optional arguments to customize the appearance of the graph:

```{r, echo = FALSE}
set.seed(7)
```

```{r}
plot(data, vertex_size = item_count, vertex_label = name, layout = "davidson-harel",
     vertex_color = item_count, vertex_color_range = c("cyan", "magenta", "green"),
     edge_color = rank, edge_color_range   = c("#555555", "#EEEEEE"), overlap_avoidance = 0.5)
```

### Subsetting

Taxonomic data can be easily subset using the **subset** function. # do something
The user can choose preserve or remove the subtaxa, supertaxa, and sequence data of the subset.
For example, **subset** can be used to look at just the Archaea:

```{r, echo = FALSE}
set.seed(1)
```

```{r}
plot(subset(data, name == "Archaea"), vertex_size = item_count, 
     vertex_label = name, vertex_color = item_count, layout = "fruchterman-reingold")
```

To make the Archaea-Bacteria division more clear, the "Root" taxon can be removed, causing two trees to be plotted:

```{r, echo = FALSE}
set.seed(2)
```

```{r}
subsetted <- subset(data, rank > 1)
plot(subsetted, vertex_size = item_count, vertex_label = name,
     vertex_color = item_count, layout = "davidson-harel", tree_label = name)
```


### Taxonomically balanced sampling


When calculating statistics for taxa, the amount of data should be balanced among taxa and there should be enough data per taxon to make unbiased estimates.
Random samples from large reference databases are biased toward overrepresented taxa.
The function **taxonomic_sample** is used to create taxonomically balanced random samples.
The acceptable range of sequence or subtaxa counts can be defined for each taxonomic rank; taxa with too few are excluded and taxa with too many are randomly sampled.
The code below samples the data such that rank 6 taxa will have 5 sequences and rank 3 taxa (phyla) will have less than 100:

```{r, echo = FALSE}
set.seed(1)
```

```{r}
sampled <- taxonomic_sample(subsetted, max_counts = c("3" = 100, "6" = 5), min_counts = c("6" = 5))
sampled <- subset(sampled, sitem_count > 0, itemless = FALSE) 
```

```{r, echo = FALSE}
set.seed(4)
```

```{r}
plot(sampled, vertex_size = item_count, vertex_label = item_count, overlap_avoidance = 0.5,
     vertex_color = item_count, layout = "davidson-harel", vertex_label_max = 100)
```


### In silico PCR

The function **primersearch** is a wrapper for the EMBOSS tool that implements *in silico* PCR.
The code below estimates the coverage of the universal bacterial primer pair 357F/519F: 

```{r}
pcr <- primersearch(sampled, forward = "CTCCTACGGGAGGCAGCAG", reverse = "GWATTACCGCGGCKGCTG",
                    pair_name = "357F_519R",  mismatch = 10)
head(taxon_data(pcr))
```

The proportion of sequences amplified can be represented by color in a metadiversity plot:

```{r, echo = FALSE}
set.seed(7) # 6, 7 is ok
```

```{r}
plot(pcr, vertex_size = item_count, vertex_label = name, vertex_color = prop_amplified,
     vertex_color_range =  c("red", "#d95f02", "#1b9e77", "#7570b3"),
     vertex_color_trans = "radius", tree_label = name)
```

This plot makes it apparent that no Archaea were amplified and most Bacteria were amplified, but not all.
The data can then be subset to better see what did not get amplified:

```{r}
library(magrittr) # Adds optional %>% operator for chaining commands
pcr %>%
  subset(name == "Bacteria") %>%
  subset(count_amplified < item_count, subtaxa = FALSE) %>% 
  plot(vertex_size = item_count, vertex_label = name, vertex_color = prop_amplified,
       vertex_color_range =  c("red", "#d95f02", "#1b9e77", "#7570b3"),
       vertex_color_interval = c(0, 1), vertex_color_trans = "radius")
```

### Differential metadiversity plots

We can compare the effectiveness of two primer pairs, 357F/519F and 515F/1100R, by plotting the difference of their amplification proportions.
First we do *in silco* PCR with 515F/1100R and combine the results:

```{r}
pcr_2 <- primersearch(sampled, forward = "GTGCCAGCMGCCGCGGTAA", reverse = "AGGGTTGCGCTCGTTG",
                      pair_name = "515F_1100R", mismatch = 10)
pcr$taxon_data$count_amplified_2 <- taxon_data(pcr_2, "count_amplified")
pcr$taxon_data$prop_diff <- taxon_data(pcr, "prop_amplified") - taxon_data(pcr_2, "prop_amplified")
```

Then, taxa that are not amplified by both pairs can be subset and the difference in amplification plotted.
In the graph below, blue means taxa were amplified by 357F/519F but not 515F/1100R; brown means the opposite:

```{r}
pcr %>%
  subset(name == "Bacteria") %>%
  subset(count_amplified < item_count | count_amplified_2 < item_count, subtaxa = FALSE) %>%
  plot(vertex_size = item_count, vertex_label = name,
       vertex_color = prop_diff, vertex_color_range = diverging_palette(),
       vertex_color_interval = c(-1, 1), vertex_color_trans = "radius")
```

### Conclusions

Intuitive plotting will reveal subtle patterns in complex data and the ability to estimate the effectiveness of new primers/barcodes will accelerate the adoption of metabarcoding to understudied groups of organisms.

### Availability

This package is currently being developed and can be installed from GitHub using the following code:

```{r, eval = FALSE}
devtools::install_github("grunwaldlab/metacoder")
```

### Aknowledgements

We thank Tom Sharpton for sharing his metagenomics expertise and advising us.
MetacodeR's major dependencies are taxize, igraph, and ggplot2.