---
title: "Comparing 16S databases"
output:
  html_document:
    toc: false
    css: custom.css
---

```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(ape)
library(metacoder)
library(knitr)
library(magrittr)
source("knitr_settings.R")
```


### SILVA

#### Parse file

The code below parses and subsets the entire SILVA non-redundant reference database.
It is done in a single step so that no large intermediate files are generated.

```{r}
set.seed(1)
file_path <- "../../../big_data/SILVA_123.1_SSURef_Nr99_tax_silva.fasta"
silva <- seqinr::read.fasta(file_path, as.string = TRUE) %>%
  extract_taxonomy(regex = "^(.*?) (.*)$",
                   key = c(id = "item_info", "class"),
                   class_sep = ";")
```


#### subset

```{r}
silva_subset <- filter_taxa(silva, taxon_ranks <= 7) %>%
  filter_taxa(name == "Bacteria", subtaxa = TRUE) %>%
  filter_taxa(item_counts >= 10)
```




#### Plot data

```{r}
filter_taxa(silva_subset, taxon_ranks < 6) %>%
  plot(vertex_size = item_counts,
       vertex_color = item_counts,
       vertex_label = name)
```


#### PCR

```{r}
silva_subset$item_data$sequence <- gsub(pattern = "u", replacement = "t",
                                        silva_subset$item_data$sequence)
pcr <- primersearch(filter_items(silva_subset),
                    forward = "CTCCTACGGGAGGCAGCAG",
                    reverse = "GWATTACCGCGGCKGCTG",
                    pair_name = "357F_519R",  mismatch = 10)
```


```{r}
plot(pcr,
     vertex_size = item_counts,
     vertex_label = name,
     vertex_color = prop_amplified,
     vertex_color_range =  c("red", "cyan"),
     vertex_color_trans = "radius")
```

```{r}
filter_taxa(pcr, prop_amplified < 0.9, supertaxa = TRUE) %>%
  plot(vertex_size = item_counts,
     vertex_label = name,
     vertex_color = prop_amplified,
     vertex_color_range =  c("red", "cyan"),
     vertex_color_trans = "radius")
```


