---
title: "Comparing 16S databases"
output:
  html_document:
    toc: false
    css: custom.css
---

```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(metacoder)
library(knitr)
library(magrittr)
source("knitr_settings.R")
```

### Analyses parameters

```{r}
input_folder <- "../../../big_data"
output_folder <- "16s_database_comparison_results"
output_format <- ".pdf"
max_taxonomy_depth <- 8
min_seq_count <- 10
max_mismatch <- 10 # percentage mismatch tolerated in pcr
pcr_success_cutoff <- 0.95 # Used to subset for graphing
```

```{r}
dir.create(output_folder, showWarnings = FALSE)
```


### SILVA

#### Parse file

The code below parses and subsets the entire SILVA non-redundant reference database.
It is done in a single step so that no large intermediate files are generated.
Even so, the object made is quite large.

```{r}
set.seed(1)
file_path <- file_path(input_folder, "SILVA_123.1_SSURef_Nr99_tax_silva.fasta")
system.time(silva <- seqinr::read.fasta(file_path, as.string = TRUE) %>%
              extract_taxonomy(regex = "^(.*?) (.*)$",
                               key = c(id = "item_info", "class"),
                               class_sep = ";") %>% 
              filter_taxa(taxon_ranks <= max_taxonomy_depth) %>%
              filter_taxa(name == "Bacteria", subtaxa = TRUE) %>%
              filter_taxa(item_counts >= min_seq_count))
```

#### Plot data

```{r}
system.time(filter_taxa(silva, taxon_ranks <= 5) %>%
              plot(vertex_size = item_counts,
                   vertex_color = item_counts,
                   vertex_label = name,
                   vertex_label_max = 150,
                   title = "SILVA sequence abundance",
                   title_size = 0.03,
                   vertex_color_axis_label = "Sequence count",
                   output_file = file.path(output_folder, paste0("silva--abundance", output_format))))
```

#### PCR

```{r}
silva$item_data$sequence <- gsub(pattern = "u", replacement = "t", silva$item_data$sequence)
system.time(pcr <- primersearch(silva,
                                forward = c("357F" = "CTCCTACGGGAGGCAGCAG"),
                                reverse = c("519R" = "GWATTACCGCGGCKGCTG"),
                                mismatch = max_mismatch))
```

```{r}
system.time(filter_taxa(pcr, taxon_ranks <= 5) %>%
              plot(vertex_size = item_counts,
                   vertex_label = name,
                   vertex_color = prop_amplified,
                   vertex_color_range =  c("red", "cyan"),
                   vertex_color_trans = "radius",
                   vertex_label_max = 150,
                   title = paste0("SILVA in silico pcr with ", pcr$item_data$pair_name[1]),
                   title_size = 0.03,
                   vertex_color_axis_label = "Proportion PCR success",
                   vertex_size_axis_label = "Sequence count",
                   output_file = file.path(output_folder, paste0("silva--pcr_all", output_format))))
```

```{r}
system.time(filter_taxa(pcr, taxon_ranks <= 5) %>%
              filter_taxa(prop_amplified < pcr_success_cutoff, supertaxa = TRUE) %>%
              plot(vertex_size = item_counts,
                   vertex_label = name,
                   vertex_color = prop_amplified,
                   vertex_color_range =  c("red", "cyan"),
                   vertex_color_trans = "radius",
                   vertex_label_max = 100,
                   title = paste0("SILVA in silico pcr with ", pcr$item_data$pair_name[1], 
                                  " ( < ", pcr_success_cutoff * 100, "%)"),
                   title_size = 0.03,
                   vertex_color_axis_label = "Proportion PCR success",
                   vertex_size_axis_label = "Sequence count",
                   output_file = file.path(output_folder, paste0("silva--pcr_less_than_", 
                                                                 pcr_success_cutoff * 100, 
                                                                 output_format))))
```

#### Cleanup

```{r}
rm(silva)
gc()
```




### RDP

```{r}
set.seed(1)
file_path <- file_path(input_folder, "rdp_current_Bacteria_unaligned.fa")
system.time(rdp <- seqinr::read.fasta(file_path, as.string = TRUE) %>% 
              extract_taxonomy(regex = "^(.*?) (.*)\\tLineage=(.*)",
                               key = c(id = "item_info", description = "item_info", "class"),
                               class_regex = "(.+?);(.*?);",
                               class_key = c("name", "taxon_info")) %>% 
              filter_taxa(taxon_ranks <= max_taxonomy_depth) %>%
              filter_taxa(name == "Bacteria", subtaxa = TRUE) %>%
              filter_taxa(item_counts >= min_seq_count))
```


#### Plot data

```{r}
system.time(filter_taxa(rdp, taxon_ranks < 6) %>%
              plot(vertex_size = item_counts,
                   vertex_color = item_counts,
                   vertex_label = name,
                   output_file = file.path(output_folder, paste0("rdp--abundance", output_format))))
```


#### PCR

```{r}
rdp$item_data$sequence <- gsub(pattern = "u", replacement = "t", rdp$item_data$sequence)
system.time(pcr <- primersearch(filter_items(rdp),
                                forward = "CTCCTACGGGAGGCAGCAG",
                                reverse = "GWATTACCGCGGCKGCTG",
                                pair_name = "357F_519R",
                                mismatch = max_mismatch))
```

```{r}
system.time(plot(pcr,
                 vertex_size = item_counts,
                 vertex_label = name,
                 vertex_color = prop_amplified,
                 vertex_color_range =  c("red", "cyan"),
                 vertex_color_trans = "radius",
                 output_file = file.path(output_folder, paste0("rdp--pcr_all", output_format))))
```

```{r}
system.time(filter_taxa(pcr, prop_amplified < 0.95, supertaxa = TRUE) %>%
              plot(vertex_size = item_counts,
                   vertex_label = name,
                   vertex_color = prop_amplified,
                   vertex_color_range =  c("red", "cyan"),
                   vertex_color_trans = "radius",
                   output_file = file.path(output_folder, paste0("rdp--pcr_less_than_",
                                                                 pcr_success_cutoff * 100,
                                                                 output_format))))
```


#### Cleanup

```{r}
rm(rdp)
gc()
```

