---
title: "Human microbiome example"
output: html_document
---

```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(knitr)
source("knitr_settings.R")
```


### Analyses parameters

```{r}
output_folder <- "human_microbiome_results"
output_format <- ".pdf"
```

### Create output directory

```{r}
dir.create(output_folder, showWarnings = FALSE)
```


### Parsing taxonomic data

The code below downloads a compressed archive and uncompresses it:

```{r}
download_and_extract <- function(url) {
  temp_dir_path <- tempdir()
  local_file_path <- file.path(temp_dir_path, basename(url))
  download.file(url = url, destfile = local_file_path, quiet = TRUE)
  unzipped_path <- R.utils::gunzip(local_file_path, overwrite = TRUE)
  unzipped_path[1]
}
```

```{r, warning=FALSE, message=FALSE}
library(metacoder)
v13_data <- parse_hmp_qiime(otu_file = "http://downloads.hmpdacc.org/data/HMQCP/otu_table_psn_v13.txt.gz",
                            mapping_file = "http://downloads.hmpdacc.org/data/HMQCP/v13_map_uniquebyPSN.txt.bz2")
v35_data <- parse_hmp_qiime(otu_file = "http://downloads.hmpdacc.org/data/HMQCP/otu_table_psn_v35.txt.gz",
                            mapping_file = "http://downloads.hmpdacc.org/data/HMQCP/v35_map_uniquebyPSN.txt.bz2")
```


### Plot of everything

```{r}
calculate_abundance <- function(data) {
  mutate_taxa(data, abundance = vapply(items(data), function(i) sum(data$item_data[i, colnames(data$item_data) %in% data$mapping$sample_id]), numeric(1)))
}

v13_data <- calculate_abundance(v13_data)
v35_data <- calculate_abundance(v35_data)


plot_all <- function(data, output_name, seed = 1) {
  set.seed(seed)
  data %>%
    filter_taxa(abundance >= 100) %>%
    plot(vertex_size = item_counts,
         vertex_size_axis_label = "Number of OTUs",
         vertex_color = abundance,
         vertex_color_trans = "area",
         vertex_color_axis_label = "Number of reads",
         vertex_label = name,
         vertex_label_max = 100,
         overlap_avoidance = 1,
         # initial_layout = "re", layout = "da",
         output_file = file.path(output_folder, paste0(output_name, "--seed_", seed, output_format)))
}


plot_all(v13_data, "hmp--v13--all_data")
plot_all(v35_data, "hmp--v35--all_data")
```




### Plot body site differences

```{r}
calculate_body_site_abundance <- function(data, site) {
  ab_col_name <- paste0(site, "_abun")
  prop_col_name <- paste0(site, "_prop")
  body_site_cols <- unlist(data$mapping[data$mapping$body_site == site, "sample_id"])
  data$taxon_data[ab_col_name] <- vapply(items(data), function(i) sum(data$item_data[i, colnames(data$item_data) %in% body_site_cols]), numeric(1))
  data$taxon_data[prop_col_name] <- data$taxon_data[[ab_col_name]] / max(data$taxon_data[[ab_col_name]])
  return(data)
}

plot_body_site_diff <- function(data, site_1, site_2, output_name, seed = 1) {
  # calculate each site data
  data <- calculate_body_site_abundance(data, site_1)
  data <- calculate_body_site_abundance(data, site_2)
   
  # filter out taxa without reads in either site
  data <- filter_taxa(data,
                      data$taxon_data[[paste0(site_1, "_prop")]] != 0 | data$taxon_data[[paste0(site_2, "_prop")]] != 0,
                      supertaxa = TRUE)
  
  # calculate scaled site difference
  prop_1 <- data$taxon_data[[paste0(site_1, "_prop")]]
  prop_2 <- data$taxon_data[[paste0(site_2, "_prop")]]
  data$taxon_data$scaled_prop_diff <- (prop_1 - prop_2) / mean(c(prop_1, prop_2))
  
  # plot
  set.seed(seed)
  data %>%
    filter_taxa(abundance >= 100) %>%
    # filter_taxa(taxon_ranks <= 5) %>%
    plot(vertex_size_axis_label = "Number of OTUs",
         vertex_size = item_counts,
         vertex_color_axis_label = paste0(site_1, " only    Both    ", site_2, " only"),
         vertex_color = scaled_prop_diff,
         vertex_color_range = diverging_palette(),
         vertex_color_trans = "area",
         vertex_color_interval = round(max(abs(range(scaled_prop_diff)))) * c(-1, 1),
         edge_color_interval = round(max(abs(range(scaled_prop_diff)))) * c(-1, 1),
         vertex_label = name,
         vertex_label_max = 150,
         overlap_avoidance = 1,
         # initial_layout = "re", layout = "da",
         # maxiter = 20, fineiter = 20,
         # margin_size = c(0.001, 0.001),
         output_file = file.path(output_folder, paste0(output_name, "--", site_1, "_vs_", site_2, "--seed_", seed, output_format)))
}

plot_body_site_diff(v13_data, "Throat", "Saliva", "hmp--v13")
plot_body_site_diff(v35_data, "Throat", "Saliva", "hmp--v35")

plot_body_site_diff(v13_data, "Supragingival_plaque", "Subgingival_plaque", "hmp--v13")
plot_body_site_diff(v35_data,  "Supragingival_plaque", "Subgingival_plaque", "hmp--v35")


plot_body_site_diff(v13_data, "Anterior_nares", "Buccal_mucosa", "hmp--v13")
plot_body_site_diff(v35_data,  "Anterior_nares", "Buccal_mucosa", "hmp--v35")

```






```{r}
plot_body_site_diff_small <- function(data, site_1, site_2, seed = 1) {
  # calculate each site data
  data <- calculate_body_site_abundance(data, site_1)
  data <- calculate_body_site_abundance(data, site_2)
   
  # filter out taxa without reads in either site
  data <- filter_taxa(data,
                      data$taxon_data[[paste0(site_1, "_prop")]] != 0 | data$taxon_data[[paste0(site_2, "_prop")]] != 0,
                      supertaxa = TRUE)
  
  # calculate scaled site difference
  prop_1 <- data$taxon_data[[paste0(site_1, "_prop")]]
  prop_2 <- data$taxon_data[[paste0(site_2, "_prop")]]
  data$taxon_data$scaled_prop_diff <- (prop_1 - prop_2) / mean(c(prop_1, prop_2))
  
  # plot
  set.seed(seed)
  data %>%
    filter_taxa(abundance >= 100) %>%
    filter_taxa(taxon_ranks <= 5) %>%
    plot(vertex_size_axis_label = "Number of OTUs",
         vertex_size = item_counts,
         vertex_color_axis_label = paste0(site_1, " only    Both    ", site_2, " only"),
         vertex_color = scaled_prop_diff,
         vertex_color_range = diverging_palette(),
         vertex_color_trans = "area",
         vertex_color_interval = round(max(abs(range(scaled_prop_diff)))) * c(-1, 1),
         edge_color_interval = round(max(abs(range(scaled_prop_diff)))) * c(-1, 1),
         vertex_label = name,
         vertex_label_max = 50,
         overlap_avoidance = 1)
         # initial_layout = "re", layout = "da",
         # maxiter = 20, fineiter = 20,
         # margin_size = c(0.001, 0.001),
         # output_file = file.path(output_folder, paste0(output_name, "--", site_1, "_vs_", site_2, "--seed_", seed, output_format)))
}


body_sites <- c("Stool", "Saliva", "Tongue_dorsum", "Buccal_mucosa")

combinations <- t(combn(seq_along(body_sites), 2))

layout_matrix <- matrix(rep(NA, (length(body_sites))^2), nrow = length(body_sites))
for (index in 1:nrow(combinations)) {
  layout_matrix[combinations[index,1], combinations[index,2]] <- index
}

bs_combinations <- t(apply(combinations, MARGIN = 1, function(x) body_sites[x]))

plots <- lapply(1:nrow(bs_combinations),
                function(i) plot_body_site_diff_small(v35_data, bs_combinations[i, 1], bs_combinations[i, 2]))
do.call(gridExtra::grid.arrange, ncol = length(body_sites) - 1, nrow =  length(body_sites) - 1,
```

