---
title: "Comparing 16S databases"
output:
  html_document:
    toc: false
    css: custom.css
---

```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(metacoder)
library(knitr)
source("knitr_settings.R")
```

## Analyses parameters

```{r}
input_folder <- "../../../big_data"
output_folder <- "16s_database_comparison_results"
output_format <- ".pdf"
max_taxonomy_depth <- 7
min_seq_count <- 10
max_mismatch <- 10 # percentage mismatch tolerated in pcr
pcr_success_cutoff <- 0.90 # Used to subset for graphing
forward = c("515F" = "GTGYCAGCMGCCGCGGTAA")
reverse = c("806R" = "GGACTACNVGGGTWTCTAAT")
```

```{r}
dir.create(output_folder, showWarnings = FALSE)
```


## SILVA

### Parse file

The code below parses and subsets the entire SILVA non-redundant reference database.
It is done in a single step so that no large intermediate files are generated.
Even so, the object made is quite large.

```{r}
set.seed(1)
file_path <- file.path(input_folder, "SILVA_123.1_SSURef_Nr99_tax_silva.fasta")
system.time(silva <- seqinr::read.fasta(file_path, as.string = TRUE) %>%
              extract_taxonomy(regex = "^(.*?) (.*)$",
                               key = c(id = "obs_info", "class"),
                               class_sep = ";") %>% 
              filter_taxa(n_supertaxa <= max_taxonomy_depth) %>%
              filter_taxa(name == "Bacteria", subtaxa = TRUE) %>%
              filter_taxa(n_obs >= min_seq_count))
```

### Plot data

```{r}
system.time(filter_taxa(silva, n_supertaxa <= 4) %>%
              plot(node_size = n_obs,
                   node_color = n_obs,
                   node_label = name,
                   node_label_max = 150,
                   title = "SILVA sequence abundance",
                   title_size = 0.03,
                   node_color_axis_label = "Sequence count",
                   output_file = file.path(output_folder, paste0("silva--abundance_rank_5", output_format))))
```

### PCR

```{r}
silva$obs_data$sequence <- gsub(pattern = "u", replacement = "t", silva$obs_data$sequence)
system.time(pcr <- primersearch(silva,
                                forward = forward_primer,
                                reverse = reverse_primer,
                                mismatch = max_mismatch))
```

```{r}
system.time(filter_taxa(pcr, n_supertaxa <= 4) %>%
              plot(node_size = n_obs,
                   node_label = name,
                   node_color = prop_amplified,
                   node_color_range =  c("red", "cyan"),
                   node_color_trans = "linear",
                   node_label_max = 150,
                   title = paste0("SILVA in silico pcr with ", pcr$obs_data$pair_name[1]),
                   title_size = 0.03,
                   node_color_axis_label = "Proportion PCR success",
                   node_size_axis_label = "Sequence count",
                   output_file = file.path(output_folder, paste0("silva--pcr_all_rank_5", output_format))))
```

```{r}
system.time(filter_taxa(pcr, n_supertaxa <= 4) %>%
              filter_taxa(prop_amplified < pcr_success_cutoff, supertaxa = TRUE) %>%
              plot(node_size = n_obs,
                   node_label = name,
                   node_color = prop_amplified,
                   node_color_range =  c("red", "cyan"),
                   node_color_trans = "linear",
                   node_label_max = 100,
                   title = paste0("SILVA in silico pcr with ", pcr$obs_data$pair_name[1], 
                                  " ( < ", pcr_success_cutoff * 100, "%)"),
                   title_size = 0.03,
                   node_color_axis_label = "Proportion PCR success",
                   node_size_axis_label = "Sequence count",
                   output_file = file.path(output_folder, paste0("silva--pcr_rank_5_less_than_", 
                                                                 pcr_success_cutoff * 100, 
                                                                 output_format))))
```




## RDP

```{r}
set.seed(1)
file_path <- file.path(input_folder, "rdp_current_Bacteria_unaligned.fa")
system.time(rdp <- seqinr::read.fasta(file_path, as.string = TRUE) %>% 
              extract_taxonomy(regex = "\\tLineage(.*)",
                               key = c("class"),
                               class_regex = "[;=](.+?);(.+?)",
                               class_key = c("name", rdp_rank = "taxon_info")) %>% 
              filter_taxa(n_supertaxa <= max_taxonomy_depth) %>%
              filter_taxa(name == "Bacteria", subtaxa = TRUE) %>%
              filter_taxa(n_obs >= min_seq_count))
```


### Plot data

```{r}
system.time(filter_taxa(rdp, n_supertaxa <= 4) %>%
              plot(node_size = n_obs,
                 node_color = n_obs,
                 node_label = name,
                 node_label_max = 150,
                 title = "RDP sequence abundance",
                 title_size = 0.03,
                 node_color_axis_label = "Sequence count",
                 output_file = file.path(output_folder, paste0("rdp--abundance_rank_5", output_format))))
```


```{r}
system.time(plot(rdp, node_size = n_obs,
                 node_color = n_obs,
                 node_label = name,
                 node_label_max = 150,
                 title = "RDP sequence abundance",
                 title_size = 0.03,
                 node_color_axis_label = "Sequence count",
                 output_file = file.path(output_folder, paste0("rdp--abundance_all_taxa", output_format))))
```


### PCR

```{r}
system.time(pcr <- primersearch(rdp,
                                forward = forward_primer,
                                reverse = reverse_primer,
                                mismatch = max_mismatch))
```

```{r}
system.time(filter_taxa(pcr, n_supertaxa <= 4) %>%
              plot(node_size = n_obs,
                 node_label = name,
                 node_color = prop_amplified,
                 node_color_range =  c("red", "cyan"),
                 node_color_trans = "linear",
                 node_label_max = 150,
                 title = paste0("RDP in silico pcr with ", unique(pcr$obs_data$pair_name)[2]),
                 title_size = 0.03,
                 node_color_axis_label = "Proportion PCR success",
                 node_size_axis_label = "Sequence count",
                 output_file = file.path(output_folder, paste0("rdp--pcr_rank_5", output_format))))
```

```{r}
system.time(plot(pcr,
                 node_size = n_obs,
                 node_label = name,
                 node_color = prop_amplified,
                 node_color_range =  c("red", "cyan"),
                 node_color_trans = "linear",
                 node_label_max = 150,
                 title = paste0("RDP in silico pcr with ", unique(pcr$obs_data$pair_name)[2]),
                 title_size = 0.03,
                 node_color_axis_label = "Proportion PCR success",
                 node_size_axis_label = "Sequence count",
                 output_file = file.path(output_folder, paste0("rdp--pcr_all_taxa", output_format))))
```



```{r}
system.time(filter_taxa(pcr, prop_amplified < pcr_success_cutoff, supertaxa = TRUE) %>%
              filter_taxa(n_supertaxa <= 4) %>%
              plot(node_size = n_obs,
                   node_label = name,
                   node_color = prop_amplified,
                   node_color_range =  c("red", "cyan"),
                   node_color_trans = "linear",
                   node_label_max = 100,
                   title = paste0("RDP in silico pcr with ", unique(pcr$obs_data$pair_name)[2], 
                                  " ( < ", pcr_success_cutoff * 100, "%)"),
                   title_size = 0.03,
                   node_color_axis_label = "Proportion PCR success",
                   node_size_axis_label = "Sequence count",
                   output_file = file.path(output_folder, paste0("rdp--pcr_rank_5_less_than_", 
                                                                 pcr_success_cutoff * 100, 
                                                                 output_format))))
```


```{r}
system.time(filter_taxa(pcr, prop_amplified < pcr_success_cutoff, supertaxa = TRUE) %>%
              plot(node_size = n_obs,
                   node_label = name,
                   node_color = prop_amplified,
                   node_color_range =  c("red", "cyan"),
                   node_color_trans = "linear",
                   node_label_max = 100,
                   title = paste0("RDP in silico pcr with ", unique(pcr$obs_data$pair_name)[2], 
                                  " ( < ", pcr_success_cutoff * 100, "%)"),
                   title_size = 0.03,
                   node_color_axis_label = "Proportion PCR success",
                   node_size_axis_label = "Sequence count",
                   output_file = file.path(output_folder, paste0("rdp--pcr_all_taxa_less_than_", 
                                                                 pcr_success_cutoff * 100, 
                                                                 output_format))))
```



## Greengenes

### Parse data

The greengenes database stores sequences in one file and taxonomy information in another and the order of the two files differ making parseing more difficult than the other databases.
Since taxonomy inforamtion is needed for creating the `taxmap` data structure, we will parse it first and add the sequence information on after.

#### Parse taxonomy file

```{r}
gg_taxonomy_path <- "data/gg_13_5_taxonomy.txt"
gg_taxonomy <- readLines(gg_taxonomy_path)
print(gg_taxonomy[1:5])
```

Note that there are some ranks with no names. 
These will be removed after parsing the file since they provide no information and an uniform-length taxonomy is not needed.

```{r}
# Parse taxonomy file
system.time(greengenes <- extract_taxonomy(input = gg_taxonomy, key = c(id = "obs_info", "class"), regex = "^([0-9]+)\t(.*)$", class_sep = "; ", class_regex = "^([a-z]{1})__(.*)$", class_key = c(rank = "taxon_info", "name")))
# Remove data for ranks with no information
greengenes <- filter_taxa(greengenes, name != "")
print(greengenes)
```


#### Parse sequence file

Next we will parse the sequence file so we can add it to the `obs_data` table of the `greengenes` object.

```{r}
gg_sequence_path <- "data/gg_13_5.fasta"
substr(readLines(gg_sequence_path, n = 10), 1, 100)
```

This can be easily parsed using `seqinr`:

```{r}
gg_sequences <- seqinr::read.fasta(gg_sequence_path, as.string = TRUE)
```


#### Integrating sequence and taxonomy

We will need to use the Greengenes ID to match up which sequence goes with which row since they are in different orders.

```{r}
greengenes <- mutate_obs(greengenes, sequence = unlist(gg_sequences)[as.character(id)])
```


### Plot data

```{r}
system.time(filter_taxa(greengenes, name == "Bacteria", subtaxa = TRUE) %>%
              plot(node_size = n_obs,
                   node_color = n_obs,
                   node_label = name,
                   node_label_max = 150,
                   title = "Greengenes sequence abundance",
                   title_size = 0.03,
                   node_color_axis_label = "Sequence count",
                   output_file = file.path(output_folder, paste0("gg--abundance_all_taxa", output_format))))
```


### PCR

```{r}
system.time(pcr <- primersearch(greengenes,
                                forward = forward,
                                reverse = reverse,
                                mismatch = max_mismatch))
```

```{r}
system.time(filter_taxa(pcr, name == "Bacteria", subtaxa = TRUE) %>%
              plot(node_size = n_obs,
                   node_label = name,
                   node_color = prop_amplified,
                   node_color_range =  c("red", "cyan"),
                   node_color_trans = "linear",
                   node_label_max = 150,
                   title = paste0("Greengenes in silico pcr with ", unique(pcr$obs_data$pair_name)[2]),
                   title_size = 0.03,
                   node_color_axis_label = "Proportion PCR success",
                   node_size_axis_label = "Sequence count",
                   output_file = file.path(output_folder, paste0("gg--pcr_all_taxa", output_format))))
```

```{r}
system.time(filter_taxa(pcr, name == "Bacteria", subtaxa = TRUE) %>%
              filter_taxa(n_obs >= 3, supertaxa = TRUE) %>%
              filter_taxa(prop_amplified < pcr_success_cutoff, supertaxa = TRUE) %>%
              plot(node_size = n_obs - count_amplified,
                   node_label = name,
                   node_color = prop_amplified,
                   node_color_range =  c("red", "orange", "yellow", "green", "cyan"),
                   node_color_trans = "linear",
                   node_color_interval = c(0, 1),
                   node_label_size_range = c(0.008, 0.015),
                   node_label_max = 500,
                   title = paste0("Greengenes in silico pcr with ", unique(pcr$obs_data$pair_name)[2], 
                                  " ( < ", pcr_success_cutoff * 100, "%)"),
                   title_size = 0.03,
                   node_color_axis_label = "Proportion PCR success",
                   node_size_axis_label = "Sequences not amplified",
                   output_file = file.path(output_folder, paste0("gg--pcr_all_taxa_less_than_", 
                                                                 pcr_success_cutoff * 100, 
                                                                 output_format))))
```

