---
title: "Introduction and workflow example"
output: 
  rmarkdown::html_vignette:
    toc: false
    css: custom.css
vignette: >
  %\VignetteIndexEntry{Introduction and workflow example}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(metacoder)
library(knitr)
opts_chunk$set(dev='svg', warning = FALSE, message = FALSE, fig.height = 7, fig.width = 7, eval = FALSE)
```

## Introduction

The goal of the `metacoder` package is to provide a set of tools for:

* Standardized parsing of taxonomic information.
* Visualization of statistics distributed over taxonomic classifications.
* Evaluating potential metabarcoding primers/loci for taxonomic specificity and discrimination.

To accomplish these goals, `metacoder` leverages resources from other R packages, interfaces with external programs, and provides novel functions where needed to allow for entire analyses within R.
Using R means that `metacoder` is entirely open-source,  works on the Linux, Windows, and Apple operating systems, and is seamlessly integrated with the best free tools for statistical analysis. 

## Workflow example

The quickest way to get idea of what the package can do is through an example of a possible workflow. 
This will briefly show the primary functions of the package, but additional details can be found in the other more focused vignettes. 

### Get sequences with assigned taxa

One of the most challenging parts of making inferences from publicly available DNA sequences for metabarcoding experiments is obtaining a set of reference sequences with reliable and specific taxonomy information. 
If the locus being studied is a standard barcoding locus and the taxa of interest well known, then there is likely a curated database to get sequences from. 
The problem becomes more difficult with non-standard loci or taxa. 
Some examples of curated databases are shown below:


Database                                        | locus         | Taxon
----------------------------------------------- | ------------- | ------------
[UNITE](https://unite.ut.ee)                    | ITS           | Fungi  
[RDP](http://rdp.cme.msu.edu/index.jsp)         | 16S, 28S rRNA | Bacteria, Archea, Fungi
[ITS1](http://itsonedb.ba.itb.cnr.it:8080/ITS1) | ITS1          | Fungi
[PR2](http://ssu-rrna.org/pr2)                  | 18S           | Protists



### Extracting taxonomy information from sequences

Surprisingly, one of the most frustrating parts of dealing with a set of sequences from diverse taxa is the assignment of a consistent taxonomy. 
Typically, curated reference databases have an included taxonomic assignment with their sequences, but they almost always have unique formats containing differing amounts of information.
The three most common sources of taxonomy information are:

* Genbank accession number
* Embedded taxonomic hierarchy
* Genbank taxon unique id

From any one of these, a hierarchy of unique taxon ids must be assigned to each sequence.
The function `extract_taxonomy` is used to parse any format containing one or more of these pieces of information.
There is another [vignette](02_extracting_taxonomy_data.html) that covers `extract_taxonomy` in more detail. 



#### Download RDP data

Lets parse the embedded taxonomic hierarchy from the RDP mothur training data set. 
R can be used to download files from the internet and decompress them.
The code below downloads the compressed data to a temporary directory.

```{r}
rdp_fasta_url <- "http://mothur.org/w/images/b/b5/Trainset10_082014.rdp.tgz"
temp_dir_path <- tempdir()
local_file_path <- file.path(temp_dir_path, basename(rdp_fasta_url))
download.file(url = rdp_fasta_url, destfile = local_file_path, quiet = TRUE)
```

#### Uncompress

Next we will uncompress the archive and identify the fasta file.

```{r}
unpacked_file_paths <- untar(local_file_path, list = TRUE) # get contents
untar(local_file_path, exdir = temp_dir_path)
unpacked_fasta_path <- file.path(temp_dir_path, 
                                  unpacked_file_paths[grepl("fasta$", unpacked_file_paths)])
```

#### Read FASTA file

We can use the `read.FASTA` function from the `ape` package to read a FASTA file and get the sequence headers. 

```{r}
sequences <- ape::read.dna(unpacked_fasta_path, format = "fasta")
```


The sequence headers are the `names` of the sequence object and have the format:

```{r, comment=NA}
cat(names(sequences)[1])
```

#### Parse sequence headers

We can now use `extract_taxonomy` to get a standardized representation of the taxonomic information.

```{r}
data <- extract_taxonomy(sequences,
                         regex = "^.*\\t(.*)",
                         key = c("class_name"),
                         class_tax_sep = ";",
                         database = "none")
```

`extract_taxonomy` returns an object of type `classfied`, which stores information assoicated with taxa as well as sequences.
We can use the `taxon_data` function to get data associated with taxa and the `item_data` function to get information associated with sequences. 

```{r}
head(taxon_data(data))
```

### Plot taxonomic information

To easily see the results, we can plot the information to see how the abundance of different taxa are distributed. 

```{r}
plot(data,
     vertex_size = item_count,
     vertex_color = item_count,
     vertex_label = name)
```

We can see that there are a lot of taxa divided between Bacteria and Archea.
Lets split the data into these two groups to make the plots easier to understand.

```{r}
data <- subset(data, rank > 1)
```

```{r}
plot(data,
     vertex_size = item_count,
     vertex_color = item_count,
     vertex_label = name,
     tree_label = name,
     layout = "davidson-harel")
```

Note that `rank` is a calculated column, meaning that its content is generated when needed.
Therefore, if we look at `rank` now, we will see that "Stamenopiles" is now rank `1` instead of `2` as it was before removeing "Eukaryota".
Note that `item_count`, `name`, and `rank` are column names in `taxon_data(data)`

```{r}
head(taxon_data(data))
```

### Subsampling

Before starting to calculate statistics for taxa within a sequence set, it is important to ensure that taxa being tested are adequately and equally represented. 
In other words, the diversity of the sequences for a given taxa should reflect the diversity of members of that taxa. 
However, there usually is not equal representation for most taxa, even in curated reference databases, since they are often disproportionately populated with culturable, model, and economically important organisms. 
Also, many reference database are very big, so sometimes it is only feasible to run computationally-intensive analyses on a subset. 

A simple random subset will not be optimal however, because it will leave out rare sequences in favor of common ones. 
To tackle this problem, `metacoder` has a function called `taxonomic_sample` that recursively travels down the taxonomy hierarchy and progressively sub-samples the database on the way back up. more information on this function can be [here](creating_a_taxonomically_balanced_reference_sequence_set.html).

The data set used for this example is small enough to use whole, but to demonstrate how `taxonomic_sample` works we will sub-sample it anyway.

```{r}
data <- taxonomic_sample(data,
                         max_counts = c("5" = 3, "6" = 1),
                         min_counts = c("3" = 3))
```

We can then plot the results as we did before, but this time lets put the item count on the vertex labels instead of the taxon name and compare the counts before and after sampling.

```{r}
plot(data,
     vertex_size = item_count,
     vertex_color = item_count,
     vertex_label = name,
     tree_label = name,
     layout = "davidson-harel")
```

Note that there are now taxa that have no sequences; these taxa did not meet the minimum out requirement and all their items were dropped, but taxon information was preserved. 
We can remove the sequenceless taxa with `subset`:

```{r}
data <- subset(data, item_count > 0, itemless = FALSE) 
```

```{r}
plot(data,
     vertex_size = item_count,
     vertex_color = item_count,
     vertex_label = name,
     tree_label = name,
     layout = "davidson-harel")
```


### Evaluating primers with *in silico* PCR

`metacoder` includes a wrapper for the *in silico* PCR program `primersearch` from the EMBOSS tool kit.
This can be used to estimate the success of different primer combinations.
The EMBOSS tool kit must be installed and in the shell's search path for the `primersearch` function to work as shown below. 

First we will do the *in silico* PCR:


```{r}
pcr_results <- primersearch(subsampled_data,
                            forward = "GCGGTAATTCCAGC",
                            reverse = "CCCGTGTTGAGTCAAAT",
                            pair_name = "F-574__and__R-1200",
                            mismatch = 10)
```

Then, we will calculate the proportion of each taxon's sequences that were able to be amplified. 
The `primersearch` function returns the item ids that amplified and we can use those ids to look up their taxon ids from the output of `extract_taxonomy`:

```{r}
plot(pcr_results,
     vertex_size = item_count,
     vertex_color = count_amplified / item_count,
     vertex_color_range = c("red", "yellow", "green"),
     vertex_label = name)
```


