<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Zachary Foster" />

<meta name="date" content="2015-03-09" />

<title>Introduction and workflow example</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title">Introduction and workflow example</h1>
<h4 class="author"><em>Zachary Foster</em></h4>
<h4 class="date"><em>2015-03-09</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#workflow-example">Workflow example</a><ul>
<li><a href="#get-sequences-with-assigned-taxa">Get sequences with assigned taxa</a></li>
<li><a href="#extracting-taxonomy-information-from-sequences">Extracting taxonomy information from sequences</a></li>
<li><a href="#subsampling">Subsampling</a></li>
<li><a href="#evaluating-primers-with-in-silico-pcr">Evaluating primers with <em>in silico</em> PCR</a></li>
<li><a href="#evaluating-primers-with-barcode-gap-analysis">Evaluating primers with barcode gap analysis</a></li>
</ul></li>
</ul>
</div>

<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>The goal of the <code>metacoder</code> package is to provide a set of tools for:</p>
<ul>
<li>Evaluating potential metabarcoding primers/loci for taxonomic specificity and discrimination.</li>
<li>Standardized parsing of taxonomic information.</li>
<li>Visualization of statistics distributed over taxonomic classifications.</li>
</ul>
<p>To accomplish these goals, <code>metacoder</code> leverages resources from other R packages, interfaces with external programs, and provides novel functions where needed to allow for entire analyses within R. Using R means that <code>metacoder</code> is entirely open-source, works on the Linux, Windows, and Apple operating systems, and is seamlessly integrated into the best free tools for statistical analysis.</p>
</div>
<div id="workflow-example" class="section level2">
<h2>Workflow example</h2>
<p>The quickest way to get idea of what the package can do is through an example of a possible workflow. This will briefly show the primary function of the package, but additional details can be found in the other more focused vignettes.</p>
<div id="get-sequences-with-assigned-taxa" class="section level3">
<h3>Get sequences with assigned taxa</h3>
<p>One of the most challenging parts of making inferences from publicly available DNA sequences for metabarcoding experiments is obtaining a set of reference sequences with reliable and specific taxonomy information. If the locus being studied is a standard barcoding locus and the taxa of interest well known, then there is likely a curated database to get sequences from. The problem becomes more difficult with non-standard loci or taxa. Some examples of curated databases are shown below:</p>
<table>
<thead>
<tr class="header">
<th align="left">Database</th>
<th align="left">locus</th>
<th align="left">Taxon</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><a href="https://unite.ut.ee">UNITE</a></td>
<td align="left">ITS</td>
<td align="left">Fungi</td>
</tr>
<tr class="even">
<td align="left"><a href="http://rdp.cme.msu.edu/index.jsp">RDP</a></td>
<td align="left">16S, 28S rRNA</td>
<td align="left">Bacteria, Archea, Fungi</td>
</tr>
<tr class="odd">
<td align="left"><a href="http://itsonedb.ba.itb.cnr.it:8080/ITS1">ITS1</a></td>
<td align="left">ITS1</td>
<td align="left">Fungi</td>
</tr>
<tr class="even">
<td align="left"><a href="http://ssu-rrna.org/pr2">PR2</a></td>
<td align="left">18S</td>
<td align="left">Protists</td>
</tr>
</tbody>
</table>
</div>
<div id="extracting-taxonomy-information-from-sequences" class="section level3">
<h3>Extracting taxonomy information from sequences</h3>
<p>Surprisingly, one of the most frustrating parts of dealing with a set of sequences from diverse taxa is the assignment of a consistent taxonomy. Typically, curated reference databases have an included taxonomic assignment with their sequences, but they almost always have unique formats containing differing amounts of information. The three most common sources of taxonomy information are:</p>
<ul>
<li>Genbank accession number</li>
<li>Embedded taxonomic hierarchy</li>
<li>Genbank taxon unique id</li>
</ul>
<p>From any one of these, a hierarchy of unique taxon ids must be assigned to each sequence. The function <code>extract_taxonomy</code> is used to parse any format containing one or more of these pieces of information. There is another <a href="extracting_taxonomy_information_from_sequence_headers.html">vignette</a> that covers <code>extract_taxonomy</code> in more detail.</p>
<pre class="sourceCode r"><code class="sourceCode r">file_path &lt;-<span class="st"> &quot;/home/local/USDA-ARS/fosterz/Downloads/gb203_pr2.fasta&quot;</span>
sequences &lt;-<span class="st"> </span>ape::<span class="kw">read.FASTA</span>(file_path)
pr2_data &lt;-<span class="st"> </span><span class="kw">extract_taxonomy</span>(<span class="kw">names</span>(sequences),
                             <span class="dt">regex =</span> <span class="st">&quot;^(.*</span><span class="ch">\\</span><span class="st">..*?)</span><span class="ch">\\</span><span class="st">|(.*)$&quot;</span>,
                             <span class="dt">key =</span> <span class="kw">c</span>(<span class="st">&quot;item_id&quot;</span>, <span class="st">&quot;class_name&quot;</span>),
                             <span class="dt">class_tax_sep =</span> <span class="st">&quot;|&quot;</span>,
                             <span class="dt">database =</span> <span class="st">&quot;none&quot;</span>)</code></pre>
</div>
<div id="subsampling" class="section level3">
<h3>Subsampling</h3>
<p>Before starting to calculate statistics for taxa within a sequence set, it is important to ensure that taxa being tested are adequately and equally represented. In other words, the diversity of the sequences for a given taxa should reflect the diversity of members of that taxa. However, there usually is not equal representation for most taxa, even in curated reference databases, since they are often disproportionately populated with culturable, model, and economically important organisms. Also, many reference database are very big, so sometimes it is only feasible to run computationally-intensive analyses on a subset. A simple random subset will not be optimal however, because it will leave out rare sequences in favor of common ones. T tackle this problem <code>metacoder</code> has a function called <code>taxonomic_sample</code> that recursively travels down the taxonomy hierarchy and progressively sub-samples the database on the way back up. more information on this function can be <a href="creating_a_taxonomically_balanced_reference_sequence_set.html">here</a>.</p>
<pre class="sourceCode r"><code class="sourceCode r">index &lt;-<span class="st"> </span><span class="kw">taxonomic_sample</span>(<span class="dt">root_id =</span> <span class="st">&quot;1&quot;</span>, <span class="dt">item_ids =</span> items$taxon_id, <span class="dt">taxon_ids =</span> taxa$taxon_id,
                          <span class="dt">parent_ids =</span> taxa$parent_id, <span class="dt">ranks =</span> taxa$rank, <span class="dt">max_counts =</span> <span class="kw">c</span>(<span class="st">&quot;o&quot;</span> =<span class="st"> </span><span class="dv">20</span>, <span class="st">&quot;s&quot;</span> =<span class="st"> </span><span class="dv">5</span>),
                          <span class="dt">min_counts =</span> <span class="kw">c</span>(<span class="st">&quot;s&quot;</span> =<span class="st"> </span><span class="dv">3</span>))
sampled_items &lt;-<span class="st"> </span>items[index, ] <span class="co"># Subsample sequence metadata</span>
sampled_taxa_index &lt;-<span class="st"> </span><span class="kw">restrict_taxonomy</span>(<span class="dt">taxa =</span> taxa$taxon_id, <span class="co"># Find which taxa are still used in the subset</span>
                                         <span class="dt">parents =</span> taxa$parent_id,
                                         <span class="dt">subset =</span> sampled_items$taxon_id)
sampled_taxa &lt;-<span class="st"> </span>taxa[sampled_taxa_index, ] <span class="co"># Subsample the taxa to only those in the subset of items </span>
sampled_taxa$item_count &lt;-<span class="st"> </span><span class="kw">get_taxon_count</span>(sampled_taxa$taxon_id, <span class="co"># Count how many items are in each taxon</span>
                                           sampled_taxa$parent_id,
                                           sampled_items$taxon_id)</code></pre>
</div>
<div id="evaluating-primers-with-in-silico-pcr" class="section level3">
<h3>Evaluating primers with <em>in silico</em> PCR</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Simulate PCR to evaluate primers -------------------------</span>
pcr &lt;-<span class="st"> </span><span class="kw">primersearch</span>(<span class="dt">sequence =</span> seq$sequence,
                    <span class="dt">forward =</span> <span class="st">&quot;TCCGTAGGTGAACCTGCGG&quot;</span>,
                    <span class="dt">reverse =</span> <span class="st">&quot;CGCATTACGTATCGCAGTTCGCAG&quot;</span>,
                    <span class="dt">seq_name =</span> taxonomy,
                    <span class="dt">pair_name =</span> <span class="st">&quot;ITS1__and__OOM-LO5.8S47B&quot;</span>,
                    <span class="dt">mismatch =</span> <span class="dv">10</span>)

<span class="co"># Calculate taxon-specific PCR success ---------------------</span>
nodes &lt;-<span class="st"> </span><span class="kw">get_nodes_from_leafs</span>(taxonomy)
seq_counts &lt;-<span class="st"> </span><span class="kw">count_nodes_in_leafs</span>(taxonomy, nodes)
pcr_counts &lt;-<span class="st"> </span><span class="kw">count_nodes_in_leafs</span>(pcr$name, nodes)
pcr_proportions &lt;-<span class="st"> </span>pcr_counts /<span class="st"> </span>seq_counts

<span class="co"># Plot results ---------------------------------------------</span>
<span class="kw">plot_value_tree</span>(nodes, pcr_proportions,
                <span class="dt">labels =</span> <span class="kw">get_tips</span>(nodes),
                <span class="dt">save =</span> <span class="st">&quot;~/test.png&quot;</span>, 
                <span class="dt">scaling =</span> seq_counts *<span class="st"> </span><span class="dv">3</span>)</code></pre>
</div>
<div id="evaluating-primers-with-barcode-gap-analysis" class="section level3">
<h3>Evaluating primers with barcode gap analysis</h3>
<pre class="sourceCode r"><code class="sourceCode r">dist_matrix &lt;-<span class="st"> </span><span class="kw">pairwise_distance</span>(<span class="dt">sequences =</span> pcr$amplicon,
                                 <span class="dt">seq_name =</span> pcr$name)


stats &lt;-<span class="st"> </span><span class="kw">calculate_barcode_statistics</span>(dist_matrix,
                                      <span class="dt">level_to_analyze=</span><span class="st">&quot;Species&quot;</span>,
                                      <span class="dt">saved_output_path =</span> <span class="st">&quot;~/test&quot;</span>,
                                      <span class="dt">save_statistics =</span> <span class="ot">TRUE</span>,
                                      <span class="dt">save_raw_data =</span> <span class="ot">TRUE</span>,
                                      <span class="dt">save_plots =</span> <span class="ot">TRUE</span>,
                                      <span class="dt">distance_bin_width =</span> <span class="fl">0.001</span>,
                                      <span class="dt">threshold_resolution =</span> <span class="fl">0.001</span>)


<span class="co">#Plot distance distribution tree graph</span>
<span class="kw">png</span>(<span class="dt">file =</span> distance_tree_plot_path, <span class="dt">bg =</span> <span class="st">&quot;transparent&quot;</span>, <span class="dt">width =</span> <span class="dv">10000</span>, <span class="dt">height =</span> <span class="dv">10000</span>)
<span class="kw">plot_image_tree</span>(stats$taxon, stats$distance_graph,
                <span class="dt">scaling=</span> stats$count, 
                <span class="dt">exclude=</span>taxa_to_exclude)
<span class="kw">dev.off</span>()

<span class="co">#Plot optimal threshold tree graph</span>
<span class="kw">png</span>(<span class="dt">file =</span> threshold_tree_plot_path, <span class="dt">bg =</span> <span class="st">&quot;transparent&quot;</span>, <span class="dt">width =</span> <span class="dv">10000</span>, <span class="dt">height =</span> <span class="dv">10000</span>)
<span class="kw">plot_image_tree</span>(stats$taxon, stats$threshold_graph,
                <span class="dt">scaling=</span> stats$count, 
                <span class="dt">exclude=</span>taxa_to_exclude)
<span class="kw">dev.off</span>()</code></pre>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
